File: v1/LLD/Design-Patterns/Decorator/code/decorator.js
/**
 * Decorator Pattern — coffee shop add-ons, stacked at runtime.
 * Run: node decorator.js
 */

class Espresso {
  getCost() { return 2.00; }
  getDescription() { return "Espresso"; }
}

class CoffeeDecorator {
  constructor(wrapped) {
    this.wrapped = wrapped;
  }
}

class Milk extends CoffeeDecorator {
  getCost() { return this.wrapped.getCost() + 0.50; }
  getDescription() { return this.wrapped.getDescription() + " + Milk"; }
}

class Caramel extends CoffeeDecorator {
  getCost() { return this.wrapped.getCost() + 0.75; }
  getDescription() { return this.wrapped.getDescription() + " + Caramel"; }
}

const plain = new Espresso();
console.log(`${plain.getDescription()} = $${plain.getCost()}`);

const fancy = new Caramel(new Milk(new Espresso()));
console.log(`${fancy.getDescription()} = $${fancy.getCost()}`);

module.exports = { Espresso, Milk, Caramel };

File: v1/LLD/Design-Patterns/Decorator/code/decorator.py
"""
Decorator Pattern — coffee shop add-ons, stacked at runtime.
Run: python decorator.py
"""

from abc import ABC, abstractmethod


class Coffee(ABC):
    @abstractmethod
    def get_cost(self):
        ...

    @abstractmethod
    def get_description(self):
        ...


class Espresso(Coffee):
    def get_cost(self):
        return 2.00

    def get_description(self):
        return "Espresso"


class CoffeeDecorator(Coffee):
    def __init__(self, wrapped: Coffee):
        self.wrapped = wrapped


class Milk(CoffeeDecorator):
    def get_cost(self):
        return self.wrapped.get_cost() + 0.50

    def get_description(self):
        return self.wrapped.get_description() + " + Milk"


class Caramel(CoffeeDecorator):
    def get_cost(self):
        return self.wrapped.get_cost() + 0.75

    def get_description(self):
        return self.wrapped.get_description() + " + Caramel"


if __name__ == "__main__":
    plain = Espresso()
    print(f"{plain.get_description()} = ${plain.get_cost()}")

    fancy = Caramel(Milk(Espresso()))
    print(f"{fancy.get_description()} = ${fancy.get_cost()}")
